bindings/src/main/java/io/pravega/storage/filesystem/FileSystemChunkStorage.java

```
@@ -0,0 +1,294 @@
+        } catch (IOException e) {
+            throw convertExeption(handle.getChunkName(), "doWrite", e);
+        }
+        return (int) totalBytesWritten;
+    }
+
+    @Override
+    public int doConcat(ConcatArgument[] chunks) throws ChunkStorageException {
+        try {
+            int totalBytesConcated = 0;
+            Path targetPath = Paths.get(config.getRoot(), chunks[0].getName());
+            long offset = chunks[0].getLength();
+
+            for (int i = 1; i < chunks.length; i++) {
+                val source = chunks[i];
+                Preconditions.checkArgument(!chunks[0].getName().equals(source.getName()), "target and source can not be same.");
+                Path sourcePath = Paths.get(config.getRoot(), source.getName());
+                long length = chunks[i].getLength();
+                Preconditions.checkState(offset <= getFileSize(targetPath));
+                Preconditions.checkState(length <= getFileSize(sourcePath));
+                try (FileChannel targetChannel = getFileChannel(targetPath, StandardOpenOption.WRITE);
+                     RandomAccessFile sourceFile = new RandomAccessFile(String.valueOf(sourcePath), "r")) {
+                    while (length > 0) {
+                        long bytesTransferred = targetChannel.transferFrom(sourceFile.getChannel(), offset, length);
+                        offset += bytesTransferred;
+                        length -= bytesTransferred;
+                    }
+                    targetChannel.force(true);
+                    totalBytesConcated += length;
+                    offset += length;
+                }
+
+            }
+            return totalBytesConcated;
+        } catch (IOException e) {
+            throw convertExeption(chunks[0].getName(), "doConcat", e);
+        }
+    }
+
+    @Override
+    protected boolean doTruncate(ChunkHandle handle, long offset) throws UnsupportedOperationException {
+        throw new UnsupportedOperationException();
```

Maybe this `throw new UnsupportedOperationException() ` should be the default implementation in the base class

---

Ok. Makes sense.

---

done